rfcs/Struct.md

```diff
@@ -0,0 +1,464 @@
+  attached (suggest we call them `STRUCT_FIELD`)
+- aliases are not allowed, otherwise the simple field merging (see below) does
+  not work (also there's no need for aliasing because struct fields are pure
+  data and do not accept arguments, so they cannot produce multiple values for
+  the same field)
+
+Both inline and named fragments are supported on `struct`s and struct unions.
+
+Field merging is very straightforward; for the example schema above, the
+following query:
+
+```graphql
+{
+  user(id: "1") {
+    bio {
+      title
+    }
+    bio {
+      socials {
+        twitter
+      }
+    }
+  }
+}
+```
+
+would be equivalent to:
+
+```graphql
+{
+  user(id: "1") {
+    bio {
+      title
+      socials {
+        twitter
+      }
+    }
+  }
+}
+```
+
+(Note that the object with both `title` and `socials{twitter}` satisfies both
+selection sets.)
+
+And similarly, the following:
+
+```graphql
+{
+  user(id: "1") {
+    ...A
+    ...B
+    ...C
+  }
+}
+
+fragment A on User {
+  bio {
+    title
+  }
+}
+
+fragment B on User {
+  bio {
+    socials {
+      twitter
+    }
+  }
+}
+
+fragment C on User {
+  # Select the entire field
+  bio
+}
+```
+
+would be equivalent to:
+
+```graphql
+{
+  user(id: "1") {
+    bio
+  }
+}
+```
+
+(Note that because we're selecting the entire `bio` field in `C`, all possible
+subselections are satisfied, and thus A and B are happy.)
+
+### Why doesn't GraphQL support wildcards over object types?
+
+In general GraphQL, wildcards have a lot of questions:
+
+- Which fields do we select?
+- What do we do for fields with arguments?
+- What do we do about deprecated fields?
+- Do we automatically add subselections to non-leaf fields?
+  - How deep do we go?
+  - What happens if there's an infinite loop?
+- Etc.
+
+They also impede the "versionless schema" desire of GraphQL - if a client uses
+wildcards then when new fields are added the client will start receiving these
+too, sending them data that they may not understand or desire and causing
+versioning problems.
+
+However, if we constrain the problem space down to that of pure structured data
+(treating it like an atom - like a `scalar` is currently) then many of these
+concerns lessen.
+
+## Should `struct` replace input objects?
+
+Maybe... They do seem to solve many of the same problems. However one major
+difference is intent; because `struct` is intended to be used for input/output
+it's intended that you pull it down, modify it as need be, and then send it back
+again. As such, adding a non-nullable field to a struct used in this way may not
+be a breaking change - existing clients should automatically receive the
+additional fields, and send them back unmodified.
+
+The semantics of `struct` and input objects are very similar, so it could be
+that we repurpose input objects for struct usage. This will need careful
+thought, discussion and investigation. Not least because the term `input` would
+make it confusing ;)
```

Comment:
another possibility - making 'struct' a new synonym of 'input', while also allowing input types to be used in outputs. This is non breaking change all the way, and we avoid introducing new concept, only an alias for existing concept.

Reply:
Yes, just using the `input` type for structs is one of the potential options. It's not clear if structs actually need to differentiate from input at all. I've used the keyword `struct` for clarity (and to make the text stand even if we decide to introduce struct as a whole new type), but I'm personally on the page of just using input objects.

Reply:
it's just if input types are allowed in the output, the term 'input' becomes confusing; so we provide 'struct' synonym, recommend it for future use, but never deprecate 'input' for backward compatibility. Struct or something else.. we'll see, but I like struct, it is a good name for the newly re-purposed input type.